### PR TITLE
Adds Union-find solution with PQ  (1101. The Earliest Moment When Everyone Become Friends)

### PR DESCRIPTION
**Pull request by altayhunoglu**
Resolves:
https://github.com/spiralgo/algorithms/issues/185

Question asks:
"**Return the earliest time for which every person became acquainted with every other person.**"

It is a textbook Union-Find question.

Therefore we can **paraphrase** the question in terms of Union-Find like this:
"`Return the earliest time for which all ids (of people) became member of the same (union, one) group.`"

We can also **rephrase** above Union-find statement like this:
"`Return the earliest time for which all ids (of people) became children of the same parent.`"

 Logs are not sorted by date.
 We can use either **PQ** or **Arrays.sort**. 
 Both have almost the same efficiency.
 
 We need to create a Log class to represent any log object, similar to what we have done in https://github.com/spiralgo/algorithms/pull/286 for Point.

### CODE REVIEW

File: src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+            return -1;
+        }
+
+        int findParent(int id) {
+            while (parent[id] != id) {
+                id = parent[id];
+            }
+            return id;
+        }
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java:55]: Is it possible to compress parents?
```suggestion
        int findParent(int id) {
            while (parent[id] != id) {
                parent[id] = parent[parent[id]];
                id = parent[id];
            }
            return id;
        }
```

Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java:55]: I have to check up on it again, but it looks that the first code already compresses the tree from the beginning.

Comment (ErdemT09) [reply src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java:55]: It shouldn't on the grounds that the parent of the children of a parent remain same when the parent of the parent is changed.
For example the pairs:     [ 1-0, 2-1, 3-2 ]  would result in the tree ( 3->2->1->0 (root-parent) )

Comment (ErdemT09) [reply src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java:55]: It shouldn't on the grounds that the parent of the children of a parent remain same when the parent of the parent is changed.
For example the pairs: [ 1-0, 2-1, 3-2, 0-4, 4-5, 5-6 ] would result in the tree ( (1,2,3)->0->4->5->6 (root-parent) )

Comment (ErdemT09) [reply src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java:55]: Compression in and of itself has a cost but it should render more efficient if such IDs are called perhaps hundreds of times.

Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java:55]: Oke, we accept the suggestion. Thanks for the feedback.

**Comment by ErdemT09** [comment]
Are there any performance differences between the two?

File: src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java (hunk 1 above)
Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/TheEarliestMomentWhenEveryoneBecomeFriends.java]: Compression takes a little bit more.
But it is up to the data population used by test cases in LeetCode.

image
image

**Comment by altayhunoglu** [comment]
> Are there any performance differences between the two?

Compression takes a little bit more.
But it is up to the data population used by test cases in LeetCode.

image
image

**Comment by ErdemT09** [comment]
There might also be a DFS solution. I personally think we can skip that approach for now, as we have done relatively enough DFS questions by now.

**Review by ErdemT09 — APPROVED** [review]

**Pull request merged**